firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,104 @@
+
+/* write parse roundtrip
+Scenario: The crc element correctly writes and parses to a buffer
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
+      # write  
+      THEN("The status of write function returns ok")
+      THEN("The crc from the crc element is equal to expected")
+      THEN("The payload of the crc element is as expected")
+      THEN("The output buffer is as expected")
+      # parse
+      THEN("The status of parse function returns ok")
+      THEN("The crc from the CRCElement is equal to the crc in the body")
+      THEN("payload from the CRCElement is equal to the expected payload from the body")
+      # write  
+      THEN("The status of write function returns ok")
+      THEN("The crc from the crc element is equal to expected")
+      THEN("The payload of the crc element is as expected")
+      THEN("The output buffer is as expected")
+
+
+/* CRCElementReceiver
+Scenario: CRCElementReceiver: The crc element receiver correctly parses datagrams into payloads
+
+  GIVEN("A CRC element receiver of capacity 254 bytes")
+    WHEN("A body with crc more than 4 bytes and equal to the crc of the payload is given to the crc_element receiver")
+      THEN("the transform method reports ok status")
+      THEN("crc returned from the crc() method is equal to the crc in the body/input_buffer")
+      THEN("payload returned from the payload() method is equal to the payload from the body")
+
+    WHEN("body with 0 bytes is given to the crc element receiver")
+      THEN("the transform status is equal to invalid parse")
+      THEN("the crc in output crc_element is equal to the given crc in input_buffer")
+
+    WHEN("body with less than 4 bytes is given to the crc element receiver")
+      // body of length 1,2,3 bytes is passed, passing 4 bytes will give a invalid_crc error instead
+      THEN("the transform status is equal to invalid parse") // for each time
+      THEN("The crc returned from the crc() method is equal to 0") // for each time
+
+    WHEN("The CRC stored in the input body is inconsistent with the payload stored in the input body")
+      THEN("the transform method reports invalid crc status")
+      THEN("the crc returned from the crc() method is equal to the invalid crc given in input body")
+
+/* CRCElementSender
+Scenario: CRCElementSender: The crc element receiver correctly generates datagrams from payload
+
+  GIVEN("A CRC element sender of capacity 254 bytes")
+    WHEN("A payload is given along with a sufficiently large output buffer")
+      THEN("the transform status is ok")
+      THEN("output buffer is as expected, equal to (crc + input_buffer)")
+
+  GIVEN("A CRC element sender of buffer size 20 bytes")
+    WHEN("The output buffer is not large enough for the input payload")
+      THEN("the transform status is equal to invalid_length")
```

Comment:
Is it possible to write a similar test for `CRCElementReceiver`? i.e. where the input buffer is too large for body_max_size

Reply:
Yes I've written a test for this, so should I then keep it in this PR and fix along with issue 282?

Reply:
Added a temporary change  for this issue in `CRCElementReceiver ` where I'm just returning an `invalid_parse ` 
if `input_buffer ` is greater than `body_max_size `, 

And for the write method I think `CRCElement::write_protected` does a resize instead to handle this issue.

Then maybe we should follow the same approach for `parse ` as well? 
If it's better to change `Vector::copy_from ` to return error codes then should it be in this PR?

Reply:
current change: 7504e9a1da4104363ef9eb80fa03677a6af17afc

Reply:
We'll include fixes for CRCElement and Datagram in this PR, and we'll reserve changes to `Vector` for a future PR

Reply:
Done with this change https://github.com/pez-globo/pufferfish-software/pull/281/commits/f947b80f5794f159f569382d81704313f06492e3